### PR TITLE
cmd/grpcurl: add -allow-unknown-fields option

### PR DESCRIPTION
**Pull request by adambabik**
This PR adds a new option `-allow-unknown-fields` (its default value is `false` so the original behavior does not change) which, if present, will instruct JSON protobuf parser to ignore unknown fields present in the request.

My use case is that I have two services sharing a message. The message got extend with a new field but the services haven't been upgraded yet. The change to the message is backward compatible but I can't use `grpcurl` due to an error: `error getting request data: message type ABC has no known field named my_new_field`.

### CODE REVIEW

**Comment by adambabik** [comment]
@jhump thanks for checking. You're totally right, in this form the package itself requires a major version bump. I will reformulate this PR to avoid that.

**Comment by adambabik** [comment]
@jhump sorry, it took me actually a while to find some spare time. I rewrote the code so that it does not alter the public interface.

File: format.go [hunk 1]
```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
 
 var _ proto.Message = (*unknownAny)(nil)
 
-// RequestParserAndFormatterFor returns a request parser and formatter for the
+// FormatterOptions is a set of flags that are passed to a JSON or text formatter.
+type FormatterOptions struct {
+	EmitJSONDefaultFields bool // FormatJSON only
+	IncludeTextSeparator  bool // FormatText only
+	AllowUnknownFields    bool // FormatJSON only
```
Comment (jhump) [inline format.go:358]: Let's group the JSON related fields together, instead of interleaving.

File: format.go [hunk 2]
```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
...
-// data (if needed by the format). The flags emitJSONDefaultFields and
-// includeTextSeparator are options for JSON and protobuf text formats,
-// respectively. Requests will be parsed from the given in.
-func RequestParserAndFormatterFor(format Format, descSource DescriptorSource, emitJSONDefaultFields, includeTextSeparator bool, in io.Reader) (RequestParser, Formatter, error) {
+// data (if needed by the format).
+// It accepts a set of options. The field EmitJSONDefaultFields and IncludeTextSeparator
+// are options for JSON and protobuf text formats, respectively. The AllowUnknownFields field
+// is a JSON-only format flag.
+// Requests will be parsed from the given in.
+func RequestParserAndFormatterWithOptions(format Format, descSource DescriptorSource, opts FormatterOptions, in io.Reader) (RequestParser, Formatter, error) {
```
Comment (jhump) [inline format.go:368]: Let's shorten this to `RequestParserAndFormatter` and then deprecate the other version.

Also, please move `opts` to be the last argument since that is more idiomatic for functions that use an options-struct pattern.

File: format.go (hunk 1 above)
Comment (jhump) [reply format.go:358]: Technically `AllowUnknownFields` is an option for the parser, not the formatter.

How about rename this to `FormatOptions` -- since they can apply to parsing the format, not just outputting it.

File: format.go [hunk 3]
```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
...
 	}
 }
 
+// RequestParserAndFormatterFor returns a request parser and formatter for the
+// given format. The given descriptor source may be used for parsing message
+// data (if needed by the format). The flags emitJSONDefaultFields and
+// includeTextSeparator are options for JSON and protobuf text formats,
+// respectively. Requests will be parsed from the given in.
+func RequestParserAndFormatterFor(format Format, descSource DescriptorSource, emitJSONDefaultFields, includeTextSeparator bool, in io.Reader) (RequestParser, Formatter, error) {
+	opts := FormatterOptions{
```
Comment (jhump) [inline format.go:387]: After moving `opts` to end of arg list for subsequent call, this could be inlined.

File: cmd/grpcurl/grpcurl.go [hunk 4]
```diff
@@ -615,7 +619,12 @@ func main() {
 				// for messages, also show a template in JSON, to make it easier to
 				// create a request to invoke an RPC
 				tmpl := grpcurl.MakeTemplate(dsc)
-				_, formatter, err := grpcurl.RequestParserAndFormatterFor(grpcurl.Format(*format), descSource, true, false, nil)
+				options := grpcurl.FormatterOptions{
+					EmitJSONDefaultFields: true,
+					IncludeTextSeparator:  false,
+					AllowUnknownFields:    *allowUnknownFields,
```
Comment (jhump) [inline cmd/grpcurl/grpcurl.go:625]: Let's just hardcode to `false`, since this code path won't be parsing anything.

Also, you can just omit the literal false values since they are default/zero vals:

```
options := grpcurl.FormatOptions{EmitJSONDefaultFields: true}
```

File: format.go [hunk 5]
```diff
@@ -342,23 +351,60 @@ func (a *unknownAny) ProtoMessage() {
...
 	}
 }
 
+// RequestParserAndFormatterFor returns a request parser and formatter for the
+// given format. The given descriptor source may be used for parsing message
+// data (if needed by the format). The flags emitJSONDefaultFields and
+// includeTextSeparator are options for JSON and protobuf text formats,
+// respectively. Requests will be parsed from the given in.
+// This function is deprecated. Please use RequestParserAndFormatter instead.
+// DEPRECATED
```
Comment (adambabik) [inline format.go:400]: I am not sure what's project's convention to mark functions as deprecated. I explained it in the comment and added `// DEPRECATED` line. The latter works nicely with editors, for example GoLand, that inform when a deprecated function is used.

**Comment by adambabik** [comment]
@jhump good stuff, thanks for a detailed review. I applied all suggestions, added a bit more code comments, and fixed tests to use `RequestParserAndFormatter` instead of now deprecated `RequestParserAndFormatterFor`.

File: format.go (hunk 5 above)
Comment (jhump) [reply format.go:400]: https://github.com/golang/go/wiki/Deprecated

This is fine. I may tweak it in a later commit. More idiomatic for Go would be to take the last two lines here and change them to something like so:

```go
// Deprecated: Use RequestParserAndFormatter instead.
```
